Add synchronous garbage collection

caesarxuchao opened the pull request:
Fix https://github.com/kubernetes/kubernetes/issues/29891.

Split into five commits:
1. generated: don't need review
2. API: got reviewed in #38678, i addressed @lavalamp's comments there.
3. registry changes: @nikhiljindal could you help take a look?
4. gc changes: reviewed by @deads2k in #38679. It needs another pass.
5. tests: @lavalamp @deads2k could take a look?

TODO:
- [ ] Update doc. Note that the existing doc has been refactored in https://github.com/kubernetes/kubernetes.github.io/pull/2488.
- [ ] add an admission controller to check if a user can set OwnerReference.BlockOwnerDeletion
- [ ] https://github.com/kubernetes/kubernetes/pull/38676#discussion_r103277274
- [ ] split the unit tests garbagecollector_test.go according to the components tested.
- [ ] try if it's practically safe to use the cached object status in attempToDeleteItem(), after synchronous GC feature is stable. (Also see https://github.com/kubernetes/kubernetes/pull/38676#discussion_r103056971)
- [ ] add blockOwnerDeletion for rs adoption https://github.com/kubernetes/kubernetes/pull/38679#discussion_r93817284
- [ ] https://k8s-gubernator.appspot.com/build/kubernetes-jenkins/pr-logs/pull/38676/pull-kubernetes-e2e-gce-etcd3/20101/ (improve the log message)

```release-note
Added foreground garbage collection: the owner object will not be deleted until all its dependents are deleted by the garbage collector. Please checkout the [user doc](https://kubernetes.io/docs/concepts/abstractions/controllers/garbage-collection/) for details.
deleteOptions.orphanDependents is going to be deprecated in 1.7. Please use deleteOptions.propagationPolicy instead.
```

k8s-reviewable commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.kubernetes.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/38676)
<!-- Reviewable:end -->

deads2k commented:
Looks like all of `pkg/controller/garbagecollector/garbagecollector.go` was replaced.  Can you open a pull with just the new files, no old changes, and I can review/merge that separately?

deads2k commented:
> Looks like all of pkg/controller/garbagecollector/garbagecollector.go was replaced. Can you open a pull with just the new files, no old changes, and I can review/merge that separately?

ah, ok.  I found the other pulls.  I understand now.

quinton-hoole commented:
@caesarxuchao Can this be closed?

caesarxuchao commented:
@quinton-hoole, I assigned it back to myself and i'll keep it open. It's split into two other PRs for reviews, but neither of them can be merged alone :)

caesarxuchao commented:
@k8s-bot unit test this

caesarxuchao commented:
@k8s-bot unit test this

caesarxuchao commented:
All tests are green.

Commit:
generated

Commit:
API

Commit:
registry changes

Commit:
gc changes

Commit:
tests

caesarxuchao commented:
@k8s-bot test this

lavalamp commented on pkg/controller/controller_ref_manager.go:
```diff
@@ -113,7 +113,7 @@ func (m *PodControllerRefManager) AdoptPod(pod *v1.Pod) error {
 	}
 	addControllerPatch := fmt.Sprintf(
-		`{"metadata":{"ownerReferences":[{"apiVersion":"%s","kind":"%s","name":"%s","uid":"%s","controller":true}],"uid":"%s"}}`,
+		`{"metadata":{"ownerReferences":[{"apiVersion":"%s","kind":"%s","name":"%s","uid":"%s","controller":true,"blockOwnerDeletion":true}],"uid":"%s"}}`,
```
What's the story for items that already exist? Do we need to do a migration?

caesarxuchao replied:
Existing controllers won't be blocked by their dependents during a foreground GC, unless the user manually patch the ownerReferences.

Doing a migration will be nice for users, or we can just communicate the fact via docs.

lavalamp commented:
huh, github published my review randomly. I'm still reviewing, not done yet.

lavalamp commented on pkg/controller/garbagecollector/graph_builder.go:
```diff
@@ -0,0 +1,500 @@
+		utilruntime.HandleError(fmt.Errorf("cannot access oldObj: %v", err))
+		return false
+	}
+	if newAccessor.GetDeletionTimestamp() == nil || oldAccessor.GetDeletionTimestamp() != nil {
```
!beingDeleted(old) && beingDeleted(new)

caesarxuchao replied:
Done. `return !beingDeleted(old) && beingDeleted(new)`

lavalamp commented on pkg/controller/garbagecollector/graph_builder.go:
```diff
@@ -0,0 +1,500 @@
+	return false
+}
+
+func hasOrphanFianlizer(accessor metav1.Object) bool {
```
s/Fian/Fina/g

caesarxuchao replied:
done.

lavalamp commented on pkg/controller/garbagecollector/graph_builder.go:
```diff
@@ -0,0 +1,500 @@
+}
+
+// if an blocking ownerReference points to an object gets removed, or gets set to
+// "BlockOwnerDeletion=false", add the object to the dirty queue.
```
s/dirty/?/

caesarxuchao replied:
s/dirty/attemptToDelete/

lavalamp reviewed (commented):
haven't gotten though the tests yet still, but here's what I've got so far.

caesarxuchao commented on staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/types.go:
```diff
@@ -73,7 +73,9 @@ type ListMeta struct {
 const (
-	FinalizerOrphan string = "orphan"
+	FinalizerOrphan           string = "orphan"
+	FinalizerOrphanDependents string = "orphan"
```
rebase error..

caesarxuchao commented on pkg/controller/replication/replication_controller.go:
```diff
@@ -488,11 +488,12 @@ func (rm *ReplicationManager) manageReplicas(filteredPods []*v1.Pod, rc *v1.Repl
+						Name:               rc.Name,
+						UID:                rc.UID,
+						BlockOwnerDeletion: &trueVar,
+						Controller:         &trueVar,
```
Ack

caesarxuchao commented on pkg/controller/controller_utils.go:
```diff
@@ -483,6 +483,9 @@ func (r RealPodControl) CreatePodsWithControllerRef(namespace string, template *
 	if controllerRef.Controller == nil || *controllerRef.Controller != true {
 		return fmt.Errorf("controllerRef.Controller is not set")
 	}
+	if controllerRef.BlockOwnerDeletion == nil || *controllerRef.BlockOwnerDeletion != true {
```
> I am confused about why we have this function

It's a util function shared by multiple controllers. We can't guarantee every author of controllers will test the controllerRef, so we make a check here.

> I am also confused about why it's so important that controllers set the block owner deletion field.

User will expect controllees to block controller's deletion during foreground deletion, so a controller should set the blockOwnerDeletion field.

caesarxuchao commented on pkg/controller/garbagecollector/garbagecollector.go:
```diff
@@ -18,237 +18,351 @@ package garbagecollector
+	glog.Infof("Garbage Collector: Initializing")
+	gc.dependencyGraphBuilder.Run(stopCh)
+	if !cache.WaitForCacheSync(stopCh, gc.dependencyGraphBuilder.HasSynced) {
+		return
```
WaitForCacheSync does the proper logging.

caesarxuchao commented on pkg/controller/garbagecollector/garbagecollector.go:
```diff
@@ -18,237 +18,351 @@ package garbagecollector
+	// gc workers
+	for i := 0; i < workers; i++ {
+		go wait.Until(gc.runAttemptToDeleteWorker, 1*time.Second, stopCh)
+		go wait.Until(gc.runAttemptToOrphanWorker, 1*time.Second, stopCh)
```
See this conversation: https://github.com/kubernetes/kubernetes/pull/38679#discussion_r92816746

caesarxuchao commented on pkg/controller/garbagecollector/garbagecollector.go:
```diff
@@ -18,237 +18,351 @@ package garbagecollector
+func (gc *GarbageCollector) classifyReferences(item *node, latestReferences []metav1.OwnerReference) (
+	solid, dangling, waitingForDependentsDeletion []metav1.OwnerReference, err error) {
+	for _, reference := range latestReferences {
+		if gc.absentOwnerCache.Has(reference.UID) {
```
Extracted to `isDangling()`.

caesarxuchao commented on pkg/controller/garbagecollector/garbagecollector.go:
```diff
@@ -258,67 +372,30 @@ func (gc *GarbageCollector) orhpanDependents(owner objectReference, dependents [
 		return fmt.Errorf("failed to orphan dependents of owner %s, got errors: %s", owner, utilerrors.NewAggregate(errorsSlice).Error())
 	}
-	glog.V(6).Infof("successfully updated all dependents")
+	glog.V(5).Infof("successfully updated all dependents")
```
Done.

caesarxuchao commented on pkg/controller/garbagecollector/garbagecollector.go:
```diff
@@ -328,485 +405,29 @@ func (gc *GarbageCollector) orphanFinalizer() {
-// GraphHasUID returns if the Propagator has a particular UID store in its
+// GraphHasUID returns if the GraphBuilder has a particular UID store in its
 // uidToNode graph. It's useful for debugging.
+// This method is used by integration tests.
```
Actually this function is safe now. `uidToNode.Read()` holds a lock. Will update the doc.

caesarxuchao commented on pkg/controller/garbagecollector/graph.go:
```diff
@@ -0,0 +1,147 @@
+	// metadata.ownerReference.
+	dependents map[*node]struct{}
+	// this is set by processEvent() if the object has non-nil DeletionTimestamp
+	// and has the FianlizerDeleteDependents.
```
Done.

caesarxuchao commented on pkg/controller/garbagecollector/graph.go:
```diff
@@ -0,0 +1,147 @@
+// blockingDependents returns the dependents that are blocking the deletion of
+// n, i.e., the dependent that has an ownerReference pointing to n, and
+// the BlockOwnerDeletion field of that ownerReference is true.
+func (n *node) blockingDependents() []*node {
```
Changed to call `getDependents()`, which is thread safe, to get a copy of the dependents at first.

lavalamp replied:
Add to the comment that the caller has no synchronization guarantees; items could get added or removed from this list the instant the function returns.

nikhiljindal commented on pkg/api/types.go:
```diff
@@ -2999,6 +3015,12 @@ type DeleteOptions struct {
+
+	// Whether and how garbage collection will be performed.
+	// Defaults to Default.
+	// Either this field or OrphanDependents may be set, but not both.
```
Add this comment to OrphanDependents as well?

caesarxuchao replied:
Done.

nikhiljindal commented on staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/types.go:
```diff
@@ -321,10 +352,17 @@ type DeleteOptions struct {
 	// +optional
 	Preconditions *Preconditions `json:"preconditions,omitempty" protobuf:"bytes,2,opt,name=preconditions"`
 
+	// Deprecated: please use the PropagationPolicy.
```
Are you sure that we want to deprecate this right away?
Shouldnt we let PropagationPolicy be there for a release, stabilize it and then deprecate OrphanDependents?

caesarxuchao replied:
I'll say it's going to be deprecated in 1.7.

caesarxuchao replied:
Done.

nikhiljindal commented on staging/src/k8s.io/apiserver/pkg/registry/generic/registry/store.go:
```diff
@@ -531,52 +531,121 @@ var (
+		switch *options.PropagationPolicy {
+		case metav1.DeletePropagationOrphan:
+			shouldOrphan = true
+		case metav1.DeletePropagationBackground, metav1.DeletePropagationForeground:
```
What about default?

caesarxuchao replied:
Default is the same as nil, so it's ignored here.

caesarxuchao replied:
Default meaning "respect the REST strategy or the existing finalizer"

caesarxuchao replied:
I removed Default from the API, since it's the same as nil.

nikhiljindal commented on staging/src/k8s.io/apiserver/pkg/registry/generic/registry/store.go:
```diff
@@ -531,52 +531,121 @@ var (
+		switch *options.PropagationPolicy {
+		case metav1.DeletePropagationForeground:
+			shouldDeleteInForeground = true
+		case metav1.DeletePropagationBackground, metav1.DeletePropagationOrphan:
```
same what about default?

nikhiljindal commented on staging/src/k8s.io/apiserver/pkg/registry/generic/registry/store.go:
```diff
@@ -531,52 +531,121 @@ var (
+		}
+	}
+
+	shouldUpdate = shouldDeleteInForeground != hasFinalizerDeleteDependents
```
optional nit: Readability will improve if you rename shouldDeleteInForeground to shouldDeleteDependents or rename hasFinalizerDeleteDependents to hasFinalizerDeleteInForeground i.e to say that the variable names should be shouldX and hasFinalizerX.

caesarxuchao replied:
Renamed the local variable `shouldDeleteInForeground` to `shouldDeleteDependentsInForeground` to be a little more consistent. The finalizer should be named a `FinalizerDeleteDependentsInForeground`, but it's too long to type everywhere.

nikhiljindal commented on staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/types.go:
```diff
@@ -73,7 +73,9 @@ type ListMeta struct {
-	FinalizerOrphan string = "orphan"
+	FinalizerOrphan           string = "orphan"
+	FinalizerOrphanDependents string = "orphan"
+	FinalizerDeleteDependents string = "foregroundDeletion"
```
Why not qualify this?
IMO, the non qualified name exception was only for existing qualifiers. We should qualify new finalizers.

lavalamp replied:
I'd lean towards consistency and leave it for now, since we already have orphan, and therefore will have to do a name change in the future anyway. It should probably have been `kubernetes.io/garbagecollector=foreground`, though. @bgrant0607 may disagree with me.

nikhiljindal commented:
I looked at the registry changes and they mostly look good to me. I have a few questions and minor suggestions.

Looking at this from the federation perspective I see that we will now have to remove FinalizerDeleteDependents as well in our federation controllers to support PropagationPolicy.Foreground.
And federation will still do foreground cascading deletion even if user sets PropagationPolicy.Background.
I am waiting for your reply on my questions to understand how will PropagationPolicy.Default work in federation.

caesarxuchao commented on pkg/controller/garbagecollector/graph.go:
```diff
@@ -0,0 +1,147 @@
+type node struct {
+	identity objectReference
+	// dependents will be read by the orphan() routine, we need to protect it with a lock.
+	dependentsLock sync.RWMutex
```
Talked in person, the concern is that if different fields are accessed in sequence using multiple locks, the caller might get an inconsistent view. There are two places these fields are accessed in sequence, but both are innocuous.
1. in `attemptToDeleteItem`,
```go
// isBeingDeleted and isDeletingDependents hold two different locks
if item.isBeingDeleted() && !item.isDeletingDependents() {
   ...
}
```
We never change beingDeleted from true to false, so it's fine.

2. also in `attemptToDeleteItem`,
```go
// acquire and then release deletingDependents lock
if item.isDeletingDependents() {
   ...
   // acquire and then release dependentsLock 
   blockingDependents := item.blockingDependents()
```
Same story here, we never change isDeletingDependents from true to false, so it's ok.

lavalamp replied:
Having requirements like that is not very future proof, please document these requirements for safe usage here.

Also, if something like being deleted is a one way trip, please change the `setBeingDeleted(bool)` function to `beginDeletion()` to make it impossible to misuse (and explain in a comment why). Likewise for deletingDependents.

caesarxuchao commented:
All comments addressed in the commit "address comments". @lavalamp @nikhiljindal PTAL. Thanks.

lavalamp commented on pkg/controller/garbagecollector/graph.go:
```diff
@@ -0,0 +1,147 @@
+	delete(ownerNode.dependents, dependent)
+}
+
+func (ownerNode *node) dependentsLength() int {
```
delete this, there is no possible way to use it safely (I think?), and it looks unused.

caesarxuchao replied:
Why it's never safe? It's used in garbagecollector.go, before GC deletes an object, it checks object.dependentsLength(), if it's 0, GC deletes the object with PropagationPolicy=background, to avoid the cycle of adding/removing finalizers.

lavalamp replied:
How could it know that nothing else is about to add a dependent? You need to lock around the add/delete object command, no?

Otherwise you can have a situation:
thread 1: (get node N) (check dependent length) (delete node)
thread 2: (get node N) (add dependent) (but node N is no longer in the graph)

lavalamp replied:
Chao explained in person that deleting means send a request to apiserver, not removing from the graph directly. I think we can probably improve this later but I guess it doesn't block the PR.

lavalamp reviewed (commented):
OK, I've now reviewed the tests.

caesarxuchao commented on pkg/controller/garbagecollector/garbagecollector.go:
```diff
@@ -163,6 +164,51 @@ func objectReferenceToMetadataOnlyObject(ref objectReference) *metaonly.Metadata
+	if err != nil {
+		return false, nil, err
+	}
+	// TODO: It's only necessary to talk to the API server if the owner node
```
Add a note in the TODO list of https://github.com/kubernetes/kubernetes/pull/38679.

caesarxuchao commented on pkg/controller/replicaset/replica_set.go:
```diff
@@ -489,14 +489,15 @@ func (rsc *ReplicaSetController) manageReplicas(filteredPods []*v1.Pod, rs *exte
 				if rsc.garbageCollectorEnabled {
-					var trueVar = true
+					var trueVar1 = true
+					var trueVar2 = true
```
Done here and other places.

caesarxuchao commented on test/e2e/framework/metrics_util.go:
```diff
@@ -117,9 +117,9 @@ var InterestingApiServerMetrics = []string{
-	"garbage_collector_orphan_processing_latency_microseconds",
+	"garbage_collector_event_queue_latency",
+	"garbage_collector_dirty_queue_latency",
+	"garbage_collector_orhan_queue_latency",
```
The name is defined by the workqueue util: https://github.com/kubernetes/kubernetes/blob/master/pkg/util/workqueue/prometheus/prometheus.go#L56-L57.

We used to set a custom metrics for garbage collector's queues, but refactored to reuse the workqueue's metrics.

caesarxuchao commented on test/e2e/garbage_collector.go:
```diff
@@ -30,8 +32,14 @@ import (
+	. "github.com/onsi/gomega"
 )
 
+func getForegroundOptions() *metav1.DeleteOptions {
```
I'll keep that in mind.

caesarxuchao commented on test/e2e/garbage_collector.go:
```diff
@@ -125,6 +135,27 @@ func verifyRemainingDeploymentsAndReplicaSets(
+			Name: name,
+		},
+		Spec: v1.PodSpec{
+			TerminationGracePeriodSeconds: func() *int64 { i := int64(0); return &i }(),
```
Changed to new(int64)

caesarxuchao commented on test/e2e/garbage_collector.go:
```diff
@@ -225,8 +256,8 @@ var _ = framework.KubeDescribe("Garbage collector", func() {
-		rc := newOwnerRC(f, rcName)
-		replicas := int32(100)
+		rc := newOwnerRC(f, rcName, 2)
+		replicas := int32(10)
```
Changed back.

caesarxuchao commented on test/e2e/garbage_collector.go:
```diff
@@ -441,4 +472,239 @@ var _ = framework.KubeDescribe("Garbage collector", func() {
+		podClient := clientSet.Core().Pods(f.Namespace.Name)
+		rc1Name := "simpletest-rc-to-be-deleted"
+		replicas := int32(100)
+		halfReplicas := 50
```
Done.

caesarxuchao commented on test/integration/replicationcontroller/replicationcontroller_test.go:
```diff
@@ -170,7 +170,7 @@ func TestAdoption(t *testing.T) {
 			},
 			func(rc *v1.ReplicationController) []metav1.OwnerReference {
-				return []metav1.OwnerReference{{UID: rc.UID, Name: rc.Name, APIVersion: "v1", Kind: "ReplicationController", Controller: &trueVar}}
+				return []metav1.OwnerReference{{UID: rc.UID, Name: rc.Name, APIVersion: "v1", Kind: "ReplicationController", Controller: &trueVar, BlockOwnerDeletion: &trueVar}}
```
Done.

caesarxuchao commented on test/integration/garbagecollector/garbage_collector_test.go:
```diff
@@ -395,19 +400,21 @@ func TestStressingCascadingDeletion(t *testing.T) {
-	count = float64(metric.Summary.GetSampleCount())
-	sum = metric.Summary.GetSampleSum()
-	t.Logf("Average time spent in GC's orphanQueue is %.1f microseconds", sum/count)
+	// TODO: figure out how to retrieve the metrics without the global
```
Done.

caesarxuchao commented on test/integration/garbagecollector/garbage_collector_test.go:
```diff
@@ -480,3 +487,183 @@ func TestOrphaning(t *testing.T) {
+	if err != nil {
+		t.Fatalf("Failed to delete the rc: %v", err)
+	}
+	time.Sleep(30 * time.Second)
```
There could be bugs causing the owner to be wrongly deleted after the suggested conditions (or other conditions) are met. We want to catch such bugs. The suggested conditions usually take just a few seconds to be met, so sleeping 30 seconds is a stronger test IMO.

lavalamp replied:
That's fair, but I'd still prefer to see it structured as "wait until condition is met; wait extra"...

Or at least it needs some comment justifying the sleep, otherwise it just looks like a flake waiting to happen. :)

You can fix in a followup.

caesarxuchao commented on test/e2e/garbage_collector.go:
```diff
@@ -441,4 +472,239 @@ var _ = framework.KubeDescribe("Garbage collector", func() {
+			for _, pod := range pods.Items {
+				framework.Logf("%#v", pod.ObjectMeta)
+			}
+			framework.Failf("failed to delete rc %s, err: %v", rc1Name, err)
```
rc1 should get deleted.

rc1 initially has 100 dependents, 50 of which also have rc2 as an owner. rc2 is not deleted. When a foreground deletion of rc1 is issued, rc1 should be deleted after the 50 pods that solely depend on rc1 are deleted. The remaining pods should only have rc2 as the owner.

This behavior is described in the [proposal](https://github.com/kubernetes/community/blob/master/contributors/design-proposals/synchronous-garbage-collection.md), starts with "if an object has multiple owners...".

lavalamp replied:
Ah, right. OK.

caesarxuchao commented on test/e2e/garbage_collector.go:
```diff
@@ -441,4 +472,239 @@ var _ = framework.KubeDescribe("Garbage collector", func() {
+		}
+		By("wait for the rc to be deleted")
+		// default client QPS is 20, deleting each pod requires 2 requests, so 30s should be enough
+		if err := wait.Poll(5*time.Second, 30*time.Second, func() (bool, error) {
```
The test is verifying that *i)* the rc is not deleted before all its 100 pods are deleted, and *ii)* that all objects are deleted within 30s.

For *i)*, the test is not strict, because the test will still pass if the rc is deleted after 99 pods are deleted, and the last pod is deleted after the rc is deleted but before the test issues `podClient.List()`. The test can be strict if it can compare the resource version of the last pod that gets deleted and the resource version of the rc when it gets deleted, but that's against the principle of "treating rv as opaque". WDYT?

> If you change the policy to background does the test reliably fail?

Yes, because the rc gets deleted immediately, and gc cannot delete all 100 pods before the `podClient.List()` is issued.

> Does it reliably pass now?

Yes so far.

lavalamp replied:
I think it's probably OK for the test to check something about RVs, since the test can know that the requirements are met. But we can fix this in a followup.

deads2k commented:
lgtm. 

It's late in the release, but I'd like to see this make 1.6.  I think its an improvement in terms of usage, reliability, and user expectation.

@lavalamp We started talking about this change last November (https://github.com/kubernetes/community/blob/master/contributors/design-proposals/synchronous-garbage-collection.md).  If you also think this an overall improvement, I think we should talk about how to get this in.

@kubernetes/sig-api-machinery-misc

caesarxuchao commented:
@lavalamp I've addressed your comments. We may have discuss more about the e2e tests. PTAL. Thanks.

lavalamp commented on pkg/controller/garbagecollector/graph.go:
```diff
@@ -67,10 +72,10 @@ func (n *node) isBeingDeleted() bool {
 }
 
-func (n *node) setDeletingDependents(v bool) {
+func (n *node) setDeletingDependentsToTrue() {
```
I would still prefer "beginDeletingDependents" or just plain "setDeletingDependents"... the "ToTrue" at the end is a bit awkward. I think just plain "set" or "mark" isn't too bad on a one-way field like this.

caesarxuchao replied:
I didn't name it "beginXXX" because that sounds like the GC is initiating the deletion of the object, while actually gc is only reflecting on the state changes received from the API server.

I'll change them to "markXXX".

caesarxuchao replied:
Done.

lavalamp commented on pkg/controller/garbagecollector/graph.go:
```diff
@@ -55,10 +55,15 @@ type node struct {
 	defer n.beingDeletedLock.Unlock()
-	n.beingDeleted = v
+	// an object is on a one way trip to its final deletion if it starts being
+	// deleted, so we only provide a function to set beingDeleted to true.
```
Put this in a godoc comment at the top of the function?

caesarxuchao replied:
Done.

lavalamp commented on pkg/controller/garbagecollector/graph.go:
```diff
@@ -55,10 +55,15 @@ type node struct {
 
-func (n *node) setBeingDeleted(v bool) {
+// WARNING: the following setters/getters use different locks, so the return
+// values of the getters can be inconsistent.
```
This is appropriately scary. Can we add it to the comment on the `node` struct instead of having it loose here?

caesarxuchao replied:
Done.

caesarxuchao commented:
Comments addressed. I'll squash the "address comments" commits after I get the lgtm. @lavalamp PTAL, thanks.

lavalamp commented:
LGTM, please squash.

caesarxuchao commented:
Applying lgtm. Squashed to 5 commits for easier reference in the future followup. Also added the release note. @lavalamp feel free to update the release note.

deads2k commented:
Bazel

caesarxuchao commented:
/lgtm

caesarxuchao commented:
@lavalamp it needs your approval.

deads2k commented:
tagged based on comments in the thread: https://github.com/kubernetes/kubernetes/pull/38676#issuecomment-282816890

deads2k commented:
and it needs another rebase. :(  Hopefully it won't be too bad.

caesarxuchao commented:
Rebased, easy enough :)

caesarxuchao commented:
Ran hack/update-staging-client.sh. Adding back the labels.

caesarxuchao commented:
@k8s-bot gci gke e2e test this

caesarxuchao commented:
Thank you all for the reviews. It's in the submit queue now.

caesarxuchao commented:
Rebased. Reapplying the lgtm.

deads2k commented:
> Automatic merge from submit-queue

How did you manage to get this comment and not merge??

deads2k commented:
Also, check the errors for https://k8s-gubernator.appspot.com/build/kubernetes-jenkins/pr-logs/pull/38676/pull-kubernetes-e2e-gce-etcd3/20101/

deads2k commented:
Based on output, it's not a bug, but you need to log the actual error you got during the verification.

```
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/garbage_collector.go:471
Feb 27 23:48:59.560: remaining rs post mortem: &v1beta1.ReplicaSetList{TypeMeta:v1.TypeMeta{Kind:"", APIVersion:""}, ListMeta:v1.ListMeta{SelfLink:"/apis/extensions/v1beta1/namespaces/e2e-tests-gc-7715r/replicasets", ResourceVersion:"12561"}, Items:[]v1beta1.ReplicaSet(nil)}
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/garbage_collector.go:451
```

You did orphan it, but you covered up the actual error you got during verification

deads2k commented:
@k8s-bot gce etcd3 e2e test this

caesarxuchao commented:
Thanks for taking care of the tests. Added to the TODO to improve the log message of that test.

caesarxuchao commented:
Rebased 3 times after got the lgtm. Adding p2 label now.
